compiler/rustc_hir_typeck/src/method/suggest.rs

```diff
@@ -3065,6 +3081,20 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 mspan,
                 format!("`{ty}` does not implement `{}`", pred.trait_ref.print_trait_sugared()),
             );
+
+            if foreign_preds.iter().any(|&(root_pred, pred)| {
+                if let ty::PredicateKind::Clause(ty::ClauseKind::Trait(root_pred)) =
+                    root_pred.kind().skip_binder()
+                    && let Some(root_adt) = root_pred.self_ty().ty_adt_def()
+                {
+                    self.tcx.is_diagnostic_item(sym::HashSet, root_adt.did())
+                        && self.tcx.is_diagnostic_item(sym::BuildHasher, pred.def_id())
+                } else {
+                    false
+                }
+            }) {
+                err.help("you might have intended to use a HashMap instead");
+            }
```

Comment:
Might be nice to extract this to a helper function that deduplicates the logic between this one and the one above, maybe named something like "`suggest_hashmap_on_unsatisfied_hashset_buildhasher`"